hack/generator/pkg/astmodel/storage_conversion_factories.go

```diff
@@ -0,0 +1,331 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package astmodel
+
+import (
+	"fmt"
+	"github.com/Azure/k8s-infra/hack/generator/pkg/astbuilder"
+	"github.com/dave/dst"
+	"github.com/pkg/errors"
+	"go/token"
+)
+
+// StorageTypeConversion generates the AST for a given conversion.
+// reader is an expression to read the original value
+// writer is an expression to write the converted value
+// Both of these might be complex expressions, possibly involving indexing into arrays or maps.
+type StorageTypeConversion func(reader dst.Expr, writer dst.Expr, generationContext *CodeGenerationContext) []dst.Stmt
+
+// StorageTypeConversionFactory represents factory methods that can be used to create StorageTypeConversions
+// for a specific pair of types
+// source is the endpoint that will be read
+// destination is the endpoint that will be written
+type StorageTypeConversionFactory func(source *StorageConversionEndpoint, destination *StorageConversionEndpoint) StorageTypeConversion
+
+// A list of all known type conversion factory methods
+var typeConversionFactories []StorageTypeConversionFactory
+
+func init() {
+	typeConversionFactories = []StorageTypeConversionFactory{
+		assignPrimitiveTypeFromPrimitiveType,
+		assignOptionalPrimitiveTypeFromPrimitiveType,
+		assignPrimitiveTypeFromOptionalPrimitiveType,
+		assignArrayFromArray,
+		assignMapFromMap,
+	}
+}
+
+// createTypeConversion tries to create a type conversion between the two provided types, using
+// all of the available type conversion functions in priority order to do so.
+func createTypeConversion(
+	sourceEndpoint *StorageConversionEndpoint,
+	destinationEndpoint *StorageConversionEndpoint) (StorageTypeConversion, error) {
+	for _, f := range typeConversionFactories {
+		result := f(sourceEndpoint, destinationEndpoint)
+		if result != nil {
+			return result, nil
+		}
+	}
+
+	err := errors.Errorf(
+		"no conversion found to assign %q from %q",
+		destinationEndpoint.name,
+		sourceEndpoint.name)
+
+	return nil, err
+}
+
+// assignPrimitiveTypeFromPrimitiveType will generate a direct assignment if both types have the
+// same underlying primitive type and have the same optionality
+//
+// <destination> = <source>
+//
+func assignPrimitiveTypeFromPrimitiveType(
+	sourceEndpoint *StorageConversionEndpoint,
+	destinationEndpoint *StorageConversionEndpoint) StorageTypeConversion {
+
+	srcOpt := AsOptionalType(sourceEndpoint.Type())
+	dstOpt := AsOptionalType(destinationEndpoint.Type())
+
+	if srcOpt != dstOpt {
```

Comment:
Shouldn't this be `if (srcOpt == nil) != (dstOpt == nil)`? Oh, I guess not since you're going to punt if they're different primitive types anyway further down. Maybe update the comment though?

Reply:
I restructure this in my next PR and it's much clearer.